apps/github-api-integration/README.md

```diff
@@ -0,0 +1,172 @@
+
+# Tutorial
+
+The three main components to this example are the Github model, the behaviours and the apps' `index.js` that configures the specfics of the app.
+
+To begin by writing the Model means when we introduce new components to the stack our upstream dependencies are known quantaties and therefore easier to reason about.
+
+## Model
+
+Our Github Model will extend from [hof-model](https://github.com/UKHomeOfficeForms/hof-model).
+```js
+const Model = require('hof-model');
+
+module.exports = class GithubModel extends Model {}
+```
+
+### Setting options at instantiation
+
+We know that [all Github API requests require a User-Agent header](https://developer.github.com/v3/#user-agent-required) so we can start by overriding the parent Model's `constructor` method to ensure all requests carry the correct header.
+
+```js
+constructor(attributes, options) {
+  // Call super first to create this.options
+  super(attributes, options);
+  // Set a custom header for the remote api
+  this.options.headers = {
+    'User-Agent': 'HOF Example API Integration'
+  };
+}
+```
+We can call `super` before setting the header to hand off the creation of `this.options` to the parent, base Model.
+
+### Setting the url
+
+The Model's `url` method will be called when any Model API method is called, but before the request to the Github API is invoked.
+In our example we override the parent Model's `url` method to return the `url` argument we passed to both the behaviours, replacing `:username` with the `name` entered by the user.
+
+```js
+url() {
+  // Augment configured url the user entered `name` param
+  // that we set on the model in the behaviour
+  return this.options.url.replace(':username', this.get('name'));
+}
+```
+
+When fetch is called from either behaviour, our Model's `url` method returns the url with which to make the request to the Github API.
+
+## Behaviours
+
+Each of our behaviours perform discreet functions.
+
+### Validate name
+The `validate-name` behaviour needs to check if the name the user has entered is a real Github username. If it is, the behaviour saves the name on the session or renders a user error if it is not.
+
+The behaviour has one method, `process`, which is called during a POST request when the form is submitted.
+
+```js
+process(req, res, next) {
+  // Set the `name` to a new instance of the Github Model
+  const name = req.form.values.name;
+  const model = new Model({name}, config);
+
+  // Make a GET request to the Github API
+  model.fetch()
+    .then(result => {
+      // the user exists, set the name on the session
+      req.sessionModel.set('name', name);
+      // continue with the request
+      super.process(req, res, next);
+    })
+    .catch(err => {
+      if (err.status === 404) {
+        // the user does not exist
+        // return an error to the user
+        return res.status(400).render('400', {
+          name: name,
+          baseUrl: req.baseUrl
+        });
+      }
```

Comment:
I would generally not handle errors in `process` but allow it to bubble through to the controller/behaviour's `errorHandler` method - https://github.com/UKHomeOfficeForms/hof-form-controller/blob/master/lib/base-controller.js#L254